### ISSUE TITLE
"Listen to letters" wav not playable on some browsers

### ISSUE THREAD

**Issue by AngelinaBelle** [SimpleMachines/SMF#1994]
correct content type is audio/vnd.wav I think. 
This link will not play on ipad safari or chrome.

http://www.rexswain.com/cgi-bin/httpview.cgi?url=http://livemods.net/21/index.php%3Faction%3Dverificationcode%3Brand%3D2da0418f45d41a031683dac0a916f22e%3Bvid%3Dregister%3Bformat%3D.wav&uag=Mozilla/5.0+(compatible%3B+MSIE+9.0%3B+Windows+NT+6.0%3B+Trident/5.0)&ref=http://www.rexswain.com/httpview.html&aen=&req=GET&ver=1.1&fmt=AUTO

![image](https://cloud.githubusercontent.com/assets/3267025/3429549/d618bb6a-0055-11e4-8d82-18507fa1195e.jpg)

**Comment by illori** [comment]
using dolphin on android, it does not play, it says it is 0:00 long. does it work with 2.0?

**Comment by AngelinaBelle** [comment]
Does not work at SimpleMachines.org either.
Or in opera at either location, for that matter -- though the direct link to it does work

**Comment by AngelinaBelle** [comment]
Can SM f force sending the correct mime type in http headers when it sends the verify audio?

See the differrence in 
http://www.rexswain.com/cgi-bin/httpview.cgi?url=http://www.kozco.com/tech/piano2.wav&uag=Mozilla/5.0+(iPad%3B+CPU+OS+7_1_1+like+Mac+OS+X)+AppleWebKit/537.51.1+(KHTML,+like+Gecko)+CriOS/35.0.1916.41+Mobile/11D201+Safari/9537.53&ref=http://www.rexswain.com/httpview.html&aen=&req=HEAD&ver=1.1&fmt=AUTO

Vs 
http://www.rexswain.com/cgi-bin/httpview.cgi?url=http://next.mmobrowser.com/index.php%3Faction%3Dverificationcode%3Brand%3Df5047aefec225a8af6ede1fb70935e19%3Bvid%3Dregister%3Bformat%3D.wav&uag=Mozilla/5.0+(iPad%3B+CPU+OS+7_1_1+like+Mac+OS+X)+AppleWebKit/537.51.1+(KHTML,+like+Gecko)+CriOS/35.0.1916.41+Mobile/11D201+Safari/9537.53&ref=http://www.rexswain.com/httpview.html&aen=&req=HEAD&ver=1.1&fmt=AUTO

Also

http://jsfiddle.net/P3uVH/#share
http://next.mmobrowser.com/index.php?topic=36.msg129#msg129

**Comment by Oldiesmann** [comment]
Where are you getting that it's sent as "image/gif"? That shouldn't happen at all with the current code as far as I can tell. We're always using "audio/x-wav" for the sound, which although not standard has been supported by many browsers for years.

Does changing "audio/x-wav" to "audio/vnd.wave" in template_verification_sound() fix it at all?

If not, what about changing to HTML5?

Find

```
        echo '
            <object type="audio/x-wav" data="', $context['verification_sound_href'], '">
                <a href="', $context['verification_sound_href'], '" rel="nofollow">', $context['verification_sound_href'], '</a>
            </object>';
```

Replace

```
        echo '
            <audio src="', $context['verification_sound_href, '" controls>
                <object type="audio/x-wav" data="', $context['verification_sound_href'], '">
                    <a href="', $context['verification_sound_href'], '" rel="nofollow">', $context['verification_sound_href'], '</a>
                </object>
            </audio>';
```

**Comment by AngelinaBelle** [comment]
The http header for the generated audio file, not the object tag in the template.
The http header is telling the browser the file is a gif rather than a wav
And some browsers will apparently refuse to treat it as a wav in that case.

When the wav file is generated, does SMF explicitly set the content type in the http header, or does it "leave it blank".
How is a URL like http://next.mmobrowser.com/index.php?action=verificationcode;rand=f5047aefec225a8af6ede1fb70935e19;vid=register;format=.wav
generated?
What content-type will it come with? Is SMF setting this explicitly, or is it leaving it up to the webserver to guess?

I imagine /x-wav would work just as well as vnd.wav
either is better than image/gif

**Comment by live627** [comment]
It explicitly sets the content-type as audio/x-wav

**Comment by AngelinaBelle** [comment]
It looks like Rex swain viewr is not the way to go (not returning session cookie?)

But still -- the issue is not going to be fixed by fixing the embedding HTML.

The DIRECT LINK (format=.wav) Does not play on the iPad in Safari or Chrome.
The ipad seems to regard something about the .wav file itself to be broken.
I don't know if this relates to the URL/file name, the header information, or the file contents

Would it be possible to have, 
a)for debug purposes, ?action=Testaudio;code=abcdef;format=.wav generate a wav file that does not depend on session data we can examine headers and contents separately?
Or b) for iPad Safari or chrome users, a message that they can listen if they switch to opera mini

**Comment by Oldiesmann** [comment]
I don't know why the direct link wouldn't work, as SMF is sending a header of "audio/x-wav" for it. If iOS' rendering engine can't interpret that properly, there really isn't a lot we can do. The reason it works on Opera Mini is because all the rendering is done on Opera's servers instead of on the client side.

**Comment by AngelinaBelle** [comment]
Hmmm....

So I downloaded the wav file (using a desktop browser that could handle that)
Interestingly, the desktop browser wants to save the file as as index.php

I upload that file to http://angelinabelle.meximas.com/test.wav

Now my ipad safari and chrome browsers have no trouble with test.wav
So perhaps the issue is not the content-type but the file name.

Perhaps the safari engine is not fooled by the fact that the URL nds in .wav.  It must realize that the file is index.php, and that .wav is just part of the query string.
Can `Content-Disposition:` be used to set a suggested filename? And will safari care?

If I have time, I will test this in an HTML document and get back to you on this.

note to self: Content-Disposition: inline; filename="myfile.txt" 
http://stackoverflow.com/questions/1741353/how-to-set-response-filename-without-forcing-saveas-dialog

**Comment by AngelinaBelle** [comment]
That doesn't seem to work either.  The iPad engine is going to refuse to render a php URL as a wav file, period, I guess.

The issue is not with the wav file itself (that works just fine).  Unless there is some other magic http header that will persuade the iPad otherwise, I think we've exhausted everything -- short of saving the wav file as an attachment, which I think maybe would be "scope creep" at this point, unless some bright new volunteer comes along to figure this out and fix every instance of CAPTCHA in SMF 2.1

This one might turn out to be a "won't fix", and that's a shame for ipad + accessibility.

http://angelinabelle.meximas.com/test.htm

```
Wave-in-a-php with content disposition: <object type="audio/x-wav" data="test2.php?format=.wav"></object>
```

```
$name = 'test.wav';


// send the right headers
header("Content-Type: audio/x-wav");
header("Content-Length: " . filesize($name));
// inline? render? attachment?
header('Content-Disposition: download; filename="', $name, '"'); // assert that this is reaaly a wav file, even though it is test2.php

// dump the picture and stop the script
$fp = fopen('./' . $name, 'rb');
fpassthru($fp);
```

**Comment by Xarcell** [comment]
I'm probably off the mark here, but have you tried changing to HTML5 markup and file format to ogg?
